### PR TITLE
Refactor page attention logic

### PR DESCRIPTION
**Pull request by liurupeng**
# Description

The changes is based on Pate and Wangyuan's work for refactoring page manager and integrating with JetStream.
It does below things:
1. Keep a global page_state and page_manager for each maxengine instance
2. Page_state will be the adapter to each prefill/insert/generate for sharing page information

tested with command: 
```
python MaxText/decode.py   MaxText/configs/base.yml   tokenizer_path=assets/tokenizer.llama2   load_parameters_path=gs://rupliu-llama2-7b-f32/unscanned/checkpoints/0/items/   max_prefill_predict_length=16   max_target_length=32   model_name=llama2-7b   ici_fsdp_parallelism=1   ici_autoregressive_parallelism=1   ici_tensor_parallelism=-1   scan_layers=false   weight_dtype=float32   per_device_batch_size=1   checkpoint_is_quantized=false  attention=paged  pagedattn_num_pages=64   pagedattn_tokens_per_page=8   pagedattn_pages_per_compute_block=4
```

The rest of the description includes relevant details and context, examples:
- why is this change being made,
- the problem being solved and any relevant context,
- why this is a good solution,
- some information about the specific implementation,
- shortcomings of the solution and possible future improvements.

If the change fixes a bug or a Github issue, please include a link, e.g.,:
FIXES: b/123456
FIXES: #123456

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/inference/page_manager.py [hunk 1]
```diff
@@ -22,6 +22,8 @@
 similar to virtual memory systems.
 """
 
+# TODO(rupliu) Add unit tests for this file
```
Comment (patemotter) [inline MaxText/inference/page_manager.py:25]: Just adding a note that these tests largely exist already, but may need some tweaking when added.

https://github.com/AI-Hypercomputer/maxtext/blob/page_manager/MaxText/tests/page_manager_test.py.

File: MaxText/layers/attentions.py [hunk 2]
```diff
@@ -1858,6 +1860,7 @@ def __call__(
       model_mode: str = common_types.MODEL_MODE_TRAIN,
       deterministic: bool = False,
       previous_chunk: Any = None,
+      slot: Optional[int] = 0,
```
Comment (patemotter) [inline MaxText/layers/attentions.py:1863]: Should we default to `None` instead of `0` in the various functions with `slot` as a parameter since `slot=0` is a valid case? It could be inadvertently hiding issues.

**Review by patemotter — COMMENTED** [review]
Overall LGTM, just a minor comment about defaults for `slot`.

File: MaxText/inference/page_manager.py (hunk 1 above)
Comment (liurupeng) [reply MaxText/inference/page_manager.py:25]: thanks, updated!

File: MaxText/layers/attentions.py (hunk 2 above)
Comment (liurupeng) [reply MaxText/layers/attentions.py:1883]: Updated

**Review by patemotter — COMMENTED** [review]
Overall LGTM, I think the only change would be to the defaults for `slot`. Right now we use a combination of 

`slot: Optional[int] = None`, 
`slot: Optional[int] = 0`, 

I think the best long-term plan is to use `slot: Optional[int] = None` everywhere and handle the logic accordingly. I have had some issues with the `0` default because it's 1) a valid slot and could be hiding issues, and 2) it could also be hiding issues because it's known at compile time and may be impacting jitted functions in a non-obvious way. 

These are based on my own experiences with debugging my version of the code, but it may not be as impactful in your version.

**Comment by liurupeng** [comment]
> Overall LGTM, I think the only change would be to the defaults for `slot`. Right now we use a combination of
> 
> `slot: Optional[int] = None`, `slot: Optional[int] = 0`,
> 
> I think the best long-term plan is to use `slot: Optional[int] = None` everywhere and handle the logic accordingly. I have had some issues with the `0` default because it's 1) a valid slot and could be hiding issues, and 2) it could also be hiding issues because it's known at compile time and may be impacting jitted functions in a non-obvious way.
> 
> These are based on my own experiences with debugging my version of the code, but it may not be as impactful in your version.

Thanks, update all default from 0 to None

**Review by patemotter — APPROVED** [review]

File: MaxText/inference/paged_attention.py [hunk 3]
```diff
@@ -331,7 +323,7 @@ def update_decode_step_pages(self, key_pages_var, value_pages_var, key, value, p
     kv_heads, num_pages, tokens_per_page, head_dim = key_pages.shape
 
     new_key = key.reshape(batch_size, kv_heads, head_dim)[:, :, :]
-    new_key = jnp.transpose(new_key, (1, 0, 2))  # [n_kv, b, d]
+    new_key = jnp.transpose(new_key, (1, 0, 2))  # [n_kv_heads, batch_size, head_dim]
```
Comment (wang2yn84) [inline MaxText/inference/paged_attention.py:326]: Nit, for comments, if you'd like to change the format from [n_kv, b, d] to [n_kv_heads, batch_size, head_dim], please keep it consistent. The comments in the following lines are still using the old format.

File: MaxText/layers/models.py [hunk 4]
```diff
@@ -164,7 +165,14 @@ class SequentialBlockDecoderLayers(nn.Module):
 
   @nn.compact
   def __call__(
-      self, inputs: jnp.ndarray, decoder_segment_ids, decoder_positions, deterministic, model_mode, page_state=None
+      self,
+      inputs: jnp.ndarray,
+      decoder_segment_ids,
+      decoder_positions,
+      deterministic,
```
Comment (wang2yn84) [inline MaxText/layers/models.py:172]: Nit, please add types.

File: MaxText/layers/models.py [hunk 5]
```diff
@@ -576,41 +588,6 @@ def setup(self):
...
-      )
-    elif model_mode == common_types.MODEL_MODE_AUTOREGRESSIVE:
-      page_state = self.page_manager(model_mode)
-    else:
-      raise ValueError(f"Unsupported model_mode {model_mode} by paged attention")
-    return page_state
-
   def __call__(
       self,
       decoder_input_tokens,
```
Comment (wang2yn84) [inline MaxText/layers/models.py:593]: Nit, type here.

File: MaxText/maxengine.py [hunk 6]
```diff
@@ -365,7 +377,7 @@ def prefill_aot(  # pylint: disable=too-many-positional-arguments
     )
 
   @functools.partial(jax.jit, static_argnums=(0,), static_argnames=("request_id",))
-  def prefill(
+  def _prefill_jit(
```
Comment (wang2yn84) [inline MaxText/maxengine.py:380]: Private? __prefill_jit?

File: MaxText/maxengine.py [hunk 7]
```diff
@@ -736,12 +793,14 @@ def prefill_insert(  # pylint: disable=too-many-positional-arguments
     return self.insert(prefix, decode_state, slot)
 
   @functools.partial(jax.jit, static_argnums=(0,), donate_argnums=(2,))
-  def generate(
+  def _generate_jit(
```
Comment (wang2yn84) [inline MaxText/maxengine.py:796]: ditto

File: MaxText/maxengine.py [hunk 8]
```diff
@@ -793,6 +853,31 @@ def generate(
...
+      params: Params,
+      decode_state: DecodeState,
+      sampler: Optional[Callable[[Any], Any]] = None,  # pylint: disable=unused-argument
+      rng: Optional[PRNGKeyType] = None,
+  ) -> Tuple[DecodeState, engine_api.ResultTokens]:
+    """Public API for generate that updates page state outside JIT."""
+    # Update page state before JIT call
+
+    if self.page_manager is not None and self.page_state is not None:
+      self.page_state = self.page_manager.reserve_decode_step_pages(self.page_state)
```
Comment (wang2yn84) [inline MaxText/maxengine.py:868]: reserve_decode_step_pages() seems jitable? Any reason why leave it outside of the jit function?

**Review by wang2yn84 — COMMENTED** [review]
Thank you Rupeng for the PR. Dropped a few comments.

File: MaxText/maxengine.py (hunk 8 above)
Comment (patemotter) [reply MaxText/maxengine.py:868]: Just tested and in the current state there doesn't seem to be any reason why we can't jit here. In other iterations there may have been a reason for omitting it, but they don't apply currently. 

`@functools.partial(jax.jit, static_argnames=("self",))` gave the same result in decode.py as without

**Review by vipannalla — APPROVED** [review]
Did a quick review, looks good. Thanks for refactoring!

File: MaxText/inference/paged_attention.py (hunk 3 above)
Comment (liurupeng) [reply MaxText/inference/paged_attention.py:326]: updated

File: MaxText/layers/models.py (hunk 4 above)
Comment (liurupeng) [reply MaxText/layers/models.py:172]: done

File: MaxText/maxengine.py (hunk 8 above)
Comment (wang2yn84) [reply MaxText/maxengine.py]: If you have any specific reason to leave it outside of the jittable function, please leave a comment here, otherwise reviewers might get confused. Thank you!

**Comment by liurupeng** [comment]
merged in copybara

File: MaxText/maxengine.py (hunk 6 above)
Comment (wang2yn84) [reply MaxText/maxengine.py:380]: Just trying to clarify, any reason for not changing it?

**Review by wyzhang — APPROVED** [review]

**Comment by liurupeng** [comment]
merged into main
